---
title: "Fraction of Optimal Codons x Protein Abundance - Ribosome Speed"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)

setwd("/stor/work/Wilke/kelly/phage_simulation")

library(tidyverse)
library(ggplot2)
library(gtools)
library(stringr)
library(cowplot)
library(readxl)
SAVE_FIGS <- TRUE

cbPalette <- c("#CC79A7", "#56B4E9", "#E69F00", "#F0E442", "#0072B2", "#D55E00", "#009E73", "#999999")

# These are genes that are not included in the simulation and should be removed before comparing
# simulations with experimental data
IGNORE_GENES <- c("gene 10B",
                "possible gene 5.5-5.7",
                "gene 1.5",
                "gene 1.6",
                "gene 4.1",
                "gene 4B",
                "gene 0.6A",
                "gene 0.6B",
                "possible gene 0.6B",
                "gene 0.5",
                "gene 0.4")
```

```{r}
labels <- read_csv("output/id_map.csv") |> filter(!is.na(gene_number))
load_sim <- function(paths) {
  data_list <- lapply(paths, function(path) {   # loop through each file path in paths
    read_tsv(path) %>%   # read the tsv file at the given path
      mutate(file = path) %>%   # add column 'file' for file path
      group_by(time) %>%   # group by time
      nest() %>%
      mutate(data = purrr::map(data, ~left_join(labels, ., by = c("pinetree" = "species")))) %>%
      unnest() %>%
      mutate(time = round(time)) %>%   # round the time column
      mutate(protein = if_else(is.na(protein), 0, protein),   # replace missing protein and transcript values with 0
             transcript = if_else(is.na(transcript), 0, transcript)) %>%
      filter(!(pinetree %in% IGNORE_GENES), !is.na(gene_number))   # filter out unwanted genes
  })
  bind_rows(data_list)   # combine data frames in data_list into a single data frame
}

load_sim <- function(paths) {
  data_list <- lapply(paths, function(path) {   # loop through each file path in paths
    read_tsv(path) %>%   # read the tsv file at the given path
      mutate(file = path) %>%   # add column 'file' for file path
      group_by(time) %>%   # group by time
      mutate(time = round(time)) %>%   # round the time column
      mutate(protein = if_else(is.na(protein), 0, protein),   # replace missing protein and transcript values with 0
             transcript = if_else(is.na(transcript), 0, as.numeric(transcript))) %>%
      filter(!(species %in% IGNORE_GENES))   # filter out unwanted genes
  })
  bind_rows(data_list)   # combine data frames in data_list into a single data frame
}
```

```{r}
max_time <- 1200
species_name <- "gene 10A"
fops <- seq(0, 0.7, by = 0.1)
pref_val <- 0.7
charge_val <- 10
ribo_speed <- 0.0008
trna_count <- 50000
dynamic_output_dir <- sprintf("data/simulation/phage/revised_dynamic_pref%.1f_charge%d_ribospeed%.4f", pref_val, charge_val, ribo_speed)

# Generate file paths
seeds <- 1:2
file_paths <- expand.grid(seed = seeds, fop = fops, pref_val = pref_val, ribo_speed = ribo_speed, trna_count = trna_count) %>%
  mutate(output_dir = dynamic_output_dir) %>%
  mutate(file_name = sprintf("trna_phage_pref%.1f_%d_fop%.1f_ribospeed%.4f_trna%d.tsv", pref_val, seed, fop, ribo_speed, trna_count)) %>%
  mutate(full_path = file.path(output_dir, file_name))

# Load all simulations
dynamic_data <- load_sim(file_paths$full_path)

# Add seed and fop columns and filter to include only the specified gene
dynamic_data <- dynamic_data %>%
  mutate(
    seed = as.numeric(str_extract(file, "(?<=_)[0-9]+(?=_fop)")),
    fop = as.numeric(str_extract(file, "(?<=fop)[0-9]+\\.[0-9]")),
  )

# Calculate average proteins of the selected species for each fop at max time
dynamic_data <- dynamic_data %>%
  filter(time == max_time, str_detect(pinetree, species_name)) %>%
  group_by(fop) %>%  
  summarize(avg_protein = mean(protein), .groups = "drop") 
```

```{r}
# Create a single plot with both models
plot <- ggplot(dynamic_data, aes(x = fop, y = avg_protein)) +
  geom_line() +
  geom_point() +
  theme_minimal() +
  labs(x = "FOP", 
       y = sprintf("%s Protein Abundance", species_name)) +
  scale_x_continuous(breaks = seq(0, 1.0, by = 0.1), labels = seq(0, 1.0, by = 0.1), trans="reverse") +
  scale_color_manual(values = cbPalette) +
  theme(legend.title = element_blank())

if (SAVE_FIGS) {
  save_plot(sprintf("analysis/figures/revised_fopXproteins_%s_ribospeed%.4f.pdf", species_name, ribo_speed), plot, base_width = 7, base_height = 5)
}
```