---
title: "Fraction of Optimal Codons x Protein Abundance"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)

setwd("/stor/work/Wilke/kelly/phage_simulation")

library(tidyverse)
library(ggplot2)
library(gtools)
library(stringr)
library(cowplot)
library(readxl)
SAVE_FIGS <- TRUE

cbPalette <- c("#CC79A7", "#56B4E9", "#E69F00", "#F0E442", "#0072B2", "#D55E00", "#009E73", "#999999")

# These are genes that are not included in the simulation and should be removed before comparing
# simulations with experimental data
IGNORE_GENES <- c("gene 10B",
                "possible gene 5.5-5.7",
                "gene 1.5",
                "gene 1.6",
                "gene 4.1",
                "gene 4B",
                "gene 0.6A",
                "gene 0.6B",
                "possible gene 0.6B",
                "gene 0.5",
                "gene 0.4")
```

```{r}
labels <- read_csv("output/id_map.csv") %>% filter(!is.na(gene_number))
load_sim <- function(paths) {
  data_list <- lapply(paths, function(path) {
    read_tsv(path) %>%
      mutate(file = path) %>%  # Add the file path as a column
      group_by(time) %>%
      nest() %>%
      mutate(data = purrr::map(data, ~left_join(labels, ., by = c("pinetree" = "species")))) %>%
      unnest() %>%
      mutate(time = round(time)) %>%
      mutate(protein = if_else(is.na(protein), 0, protein),
             transcript = if_else(is.na(transcript), 0, transcript)) %>%
      filter(!(pinetree %in% IGNORE_GENES), !is.na(gene_number))
  })
  bind_rows(data_list)
}

# Define the output directory
output_dir <- "data/simulation/phage/codon_parallel_pref0.7_charge100"
max_time <- 825
gene_name <- "11"

# Extract pref_val and charge_val from the output directory string
pref_val <- as.numeric(str_extract(output_dir, "(?<=pref)[0-9]\\.[0-9]"))
charge_val <- as.numeric(str_extract(output_dir, "(?<=charge)[0-9]+"))

# Generate file paths
seeds <- 1:3
fop_proportions <- 1:9

file_paths <- expand.grid(seed = seeds, fop = fop_proportions) %>%
  mutate(file_name = sprintf("trna_phage_pref%.1f_%d_fop%d.tsv", pref_val, seed, fop)) %>%
  mutate(full_path = file.path(output_dir, file_name))

# Load all simulations
all_data <- load_sim(file_paths$full_path)

# Add seed, fop columns
  data <- data %>%
    mutate(
      seed = as.numeric(str_extract(file, "(?<=_)[0-9]+(?=_fop)")),
      fop = as.numeric(str_extract(file, "(?<=fop)[0-9]+")),
    )

# Filter the data to include only the specified gene
gene_data <- all_data %>%
  filter(str_detect(pinetree, paste("gene", gene_name)))

# Calculate average proteins for each proportion at max time
avg_proteins <- gene_data %>%
  filter(time == max_time) %>%
  group_by(fop) %>%
  summarize(avg_protein = mean(protein), .groups = "drop")
print(avg_proteins)

# Create the plot
avg_plot <- ggplot(avg_proteins, aes(x = fop, y = avg_protein)) +
  geom_line() +
  geom_point() +
  theme_minimal() +
  labs(x = "Fraction of Optimal Codons", 
    y = paste("Average Proteins at", max_time, "s for gene", gene_name), 
    title = paste("Average Proteins at", max_time, "s for gene", gene_name, "for Fractions of Optimal Codons   [ tRNA pref proportion:", pref_val, "& charge rate:", charge_val,"]")) +
  scale_x_continuous(breaks = seq(1, 9, by = 1), labels = seq(0.1, 0.9, by = 0.1))

if (SAVE_FIGS) {
  save_plot(sprintf("analysis/figures/avg_gene%sproteins_by_codon_proportion_%ds_pref%.1f_charge%d.pdf", gene_name, max_time, pref_val, charge_val), avg_plot, base_width = 12, base_height = 12)
}
```