---
title: "Average Overall FOP Over Time"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)

setwd("/stor/work/Wilke/kelly/phage_simulation")

library(tidyverse)
library(ggplot2)
library(gtools)
library(stringr)
library(cowplot)
library(readxl)
SAVE_FIGS <- TRUE

cbPalette <- c("#CC79A7", "#56B4E9", "#E69F00", "#F0E442", "#0072B2", "#D55E00", "#009E73", "#999999")

# These are genes that are not included in the simulation and should be removed before comparing
# simulations with experimental data
IGNORE_GENES <- c("gene 10B",
                "possible gene 5.5-5.7",
                "gene 1.5",
                "gene 1.6",
                "gene 4.1",
                "gene 4B",
                "gene 0.6A",
                "gene 0.6B",
                "possible gene 0.6B",
                "gene 0.5",
                "gene 0.4")
```

```{r}
labels <- read_csv("output/id_map.csv") %>% filter(!is.na(gene_number))
load_sim <- function(paths) {
  data_list <- lapply(paths, function(path) {
    read_tsv(path) %>%
      mutate(file = path) %>%
      group_by(time) %>%
      nest() %>%
      mutate(data = purrr::map(data, ~full_join(labels, ., by = c("pinetree" = "species")))) %>%
      unnest() %>%
      mutate(time = round(time)) %>%
      mutate(protein = if_else(is.na(protein), 0, protein),
             transcript = if_else(is.na(transcript), 0, transcript)) %>%
      filter(!(pinetree %in% IGNORE_GENES), !is.na(gene_number))
  })
  bind_rows(data_list)
}

# Define the output directories and max time
output_dirs <- c("data/simulation/phage/codon_parallel_pref0.7_charge10",
                 "data/simulation/phage/codon_parallel_pref0.5_charge10")
max_time <- 825

# Extract pref_val and charge_val from the output directory string
pref_vals <- as.numeric(str_extract(output_dirs, "(?<=pref)[0-9]\\.[0-9]"))
charge_val <- as.numeric(str_extract(output_dirs[1], "(?<=charge)[0-9]+"))

# Generate file paths for each charge rate
seeds <- 1:3
fop_proportions <- 0:10

file_paths <- expand.grid(seed = seeds, fop = fop_proportions, pref_val = pref_vals) %>%
  mutate(output_dir = ifelse(pref_val == 0.7, output_dirs[1], output_dirs[2])) %>%
  mutate(file_name = sprintf("trna_phage_pref%.1f_%d_fop%d.tsv", pref_val, seed, fop)) %>%
  mutate(full_path = file.path(output_dir, file_name))

# Load the FOP scores and lengths
fops_lengths <- read_csv("src/python/models/trna_phage_model/fop_scores_and_lengths.csv") 

# Load simulations and join with FOP scores and lengths
all_data <- load_sim(file_paths$full_path) %>%
  left_join(fops_lengths, by = c("pinetree" = "gene_name")) 

# Add seed, fop, pref columns
all_data <- all_data %>%
  mutate(
    seed = as.numeric(str_extract(file, "(?<=_)[0-9]+(?=_fop)")),
    fop = as.numeric(str_extract(file, "(?<=fop)[0-9]+")),
    pref = as.numeric(str_extract(file, "(?<=pref)[0-9]\\.[0-9]"))
  )
```

```{r}
# Add columns calculating total number of optimal codons and all codons in each gene
all_data <- all_data %>%
  mutate(
    total_optimal_codons = length * transcript * fop_score,
    total_codons = length * transcript
  )

# Calculate total optimal codons and total codons across all genes for each time point
all_data_summary <- all_data %>%
  group_by(time, pref) %>%
  summarise(
    overall_total_optimal_codons = sum(total_optimal_codons, na.rm = TRUE),
    overall_total_codons = sum(total_codons, na.rm = TRUE)
  ) %>%
  mutate(avg_overall_fop = overall_total_optimal_codons / overall_total_codons)

lineplot <- ggplot(na.omit(all_data_summary), aes(x = time, y = avg_overall_fop, color = as.factor(pref))) +
  geom_line() +
  theme_minimal() +
  labs(x = "Time (s)", 
    y = paste("Average Overall Fraction of Optimal Codons"), 
    title = paste("Average Overall Fraction of Optimal Codons Over Time\n[ tRNA Charge Rate:", charge_val, "]"),
    color = "Fraction of Preferred tRNAs") +
  scale_color_manual(values = cbPalette) 

if (SAVE_FIGS) {
  save_plot(sprintf("analysis/figures/avg_overall_fop_over_time_multipref_charge%d.pdf", charge_val), lineplot, base_width = 6, base_height = 6)
}
```