---
title: "Codon Model Analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)

setwd("/stor/work/Wilke/kelly/phage_simulation")

library(tidyverse)
library(ggplot2)
library(gtools)
library(stringr)
library(cowplot)
library(readxl)
SAVE_FIGS <- TRUE

cbPalette <- c("#CC79A7", "#56B4E9", "#E69F00", "#F0E442", "#0072B2", "#D55E00", "#009E73", "#999999")

# These are genes that are not included in the simulation and should be removed before comparing
# simulations with experimental data
IGNORE_GENES <- c("gene 10B",
                "possible gene 5.5-5.7",
                "gene 1.5",
                "gene 1.6",
                "gene 4.1",
                "gene 4B",
                "gene 0.6A",
                "gene 0.6B",
                "possible gene 0.6B",
                "gene 0.5",
                "gene 0.4")
```

```{r}
id_map <- read_csv("output/id_map.csv") %>% filter(!is.na(gene_number), !(pinetree %in% IGNORE_GENES))
gene_labels <- unique(mixedsort(id_map$gene_number))
```

```{r}
labels <- read_csv("output/id_map.csv") %>% filter(!is.na(gene_number))
load_sim <- function(paths) {
  data_list <- lapply(paths, function(path) {
    read_tsv(path) %>%
      mutate(file = path) %>%  # Add the file path as a column
      group_by(time) %>%
      nest() %>%
      mutate(data = purrr::map(data, ~left_join(labels, ., by = c("pinetree" = "species")))) %>%
      unnest() %>%
      mutate(time = round(time)) %>%
      mutate(protein = if_else(is.na(protein), 0, protein),
             transcript = if_else(is.na(transcript), 0, transcript)) %>%
      filter(!(pinetree %in% IGNORE_GENES), !is.na(gene_number))
  })
  bind_rows(data_list)
}

# Define the output directory
output_dir <- "data/simulation/phage/codon_parallel_output"

# Generate file paths
seeds <- 1:3
fop_proportions <- 1:9

file_paths <- expand.grid(seed = seeds, fop = fop_proportions) %>%
  mutate(file_name = sprintf("trna_phage_pref0.7_%d_fop%d.tsv", seed, fop)) %>%
  mutate(full_path = file.path(output_dir, file_name))

# Load all simulations
all_data <- load_sim(file_paths$full_path)

# Add proportion and seed columns
all_data <- all_data %>%
  mutate(
    seed = as.numeric(str_extract(file, "(?<=_)[0-9](?=_fop)")),
    proportion = as.numeric(str_extract(file, "(?<=pref)[0-9]\\.[0-9]")),
    fop = as.numeric(str_extract(file, "(?<=fop)[0-9]+"))
  )

# Filter the data to include only gene 10A
gene_10A_data <- all_data %>%
  filter(str_detect(pinetree, "gene 10A"))

# Calculate average proteins for each proportion at 420 seconds
avg_proteins_825 <- gene_10A_data %>%
  filter(time == 825) %>%
  group_by(fop) %>%
  summarize(avg_protein = mean(protein), .groups = "drop")
print(avg_proteins_825)

# Create the plot
avg_plot_825 <- ggplot(avg_proteins_825, aes(x = fop, y = avg_protein)) +
  geom_line() +
  geom_point() +
  theme_minimal() +
  labs(x = "Proportion", y = "Average Proteins at 825s for gene 10A") +
  scale_x_continuous(breaks = seq(1, 9, by = 1))


if (SAVE_FIGS) {
  save_plot("analysis/figures/codon_avg_gene10proteins_by_proportion_825s.pdf", avg_plot_825, base_width = 12, base_height = 16)
}
```