---
title: "Old Model - Total Protein Abundance by Weight Factor"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)

setwd("/stor/work/Wilke/kelly/phage_simulation")

library(tidyverse)
library(ggplot2)
library(gtools)
library(stringr)
library(cowplot)
library(readxl)
SAVE_FIGS <- TRUE

cbPalette <- c("#CC79A7", "#56B4E9", "#E69F00", "#F0E442", "#0072B2", "#D55E00", "#009E73", "#999999")

# These are genes that are not included in the simulation and should be removed before comparing
# simulations with experimental data
IGNORE_GENES <- c("gene 10B",
                "possible gene 5.5-5.7",
                "gene 1.5",
                "gene 1.6",
                "gene 4.1",
                "gene 4B",
                "gene 0.6A",
                "gene 0.6B",
                "possible gene 0.6B",
                "gene 0.5",
                "gene 0.4")
```

```{r}
# Load labels
labels <- read_csv("output/id_map.csv") |> filter(!is.na(gene_number))
# Function to load and process simulation data
load_sim <- function(paths) {
  data_list <- lapply(paths, function(path) {
    read_tsv(path) %>%
      mutate(file = path) %>%  # Add the file path as a column
      group_by(time) %>%
      nest() %>%
      mutate(data = purrr::map(data, ~left_join(labels, ., by = c("pinetree" = "species")))) %>%
      unnest() %>%
      mutate(time = round(time)) %>%
      mutate(protein = if_else(is.na(protein), 0, protein),
             transcript = if_else(is.na(transcript), 0, transcript)) %>%
      filter(!(pinetree %in% IGNORE_GENES), !is.na(gene_number))
  })
  bind_rows(data_list)
}
```

```{r}
max_time <- 1200
model_output_dir <- sprintf("data/simulation/phage/phage_model_fixed")

# Define the multipliers with specific values
multipliers <- c(0.001, 0.01, seq(0.1, 1.0, by = 0.1), 10.0, 100.0)

# Generate file paths
seed_vals <- 1:3
file_paths <- expand.grid(multiplier = multiplier_vals, seed = seed_vals) %>%
  mutate(output_dir = model_output_dir) %>%
  mutate(file_name = sprintf("phage_model_fixed_weight_factor%.3f_seed%d.tsv", multiplier, seed)) %>%
  mutate(full_path = file.path(output_dir, file_name))

# Load all simulations
sim_data <- load_sim(file_paths$full_path)

# Add multiplier and seed columns
sim_data <- sim_data %>%
  mutate(
    multiplier =  as.numeric(str_extract(file, "(?<=weight_factor)[0-9]{1,3}\\.[0-9]+(?=_seed)")),
    seed = as.numeric(str_extract(file, "(?<=_seed)[0-9]+(?=\\.tsv)"))
  )

# Calculate average sum of proteins for each multiplier at max time
sims <- sim_data %>%
  filter(time == max_time) %>%
  group_by(multiplier, seed) %>%
  summarize(sum_protein = sum(protein, na.rm = TRUE)) %>%
  group_by(multiplier) %>%
  summarize(avg_protein = mean(sum_protein, na.rm = TRUE))

# Remove rows with NA multiplier
sims <- sims %>%
  filter(!is.na(multiplier))

# Adjust multiplier to ensure that 1 maps to 0 on the log scale
sims <- sims %>%
  mutate(multiplier_log = log2(multiplier))  # Apply log2 transformation directly to the multiplier

# Create the plot with the adjusted log scale
plot <- ggplot(sims, aes(x = multiplier_log, y = avg_protein)) +
  geom_line() +
  geom_point() +
  theme_minimal() +
  labs(x = "Log2(Weight Factor)", 
       y = paste("Total Protein Abundance at", max_time, "s"), 
       title = paste("Total Protein Abundance at", max_time, "s\nfor Fractions of Optimal Codons   [ Using the Old Model ]"))

if (SAVE_FIGS) {
  save_plot(sprintf("analysis/figures/fixed_total_proteins_by_weight_factor_%ds.pdf", max_time), plot, base_width = 6, base_height = 6)
}
```