---
title: "Old Model - Total Protein Abundance by Weight Factor"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)

setwd("/stor/work/Wilke/kelly/phage_simulation")

library(tidyverse)
library(ggplot2)
library(gtools)
library(stringr)
library(cowplot)
library(readxl)
SAVE_FIGS <- TRUE

cbPalette <- c("#CC79A7", "#56B4E9", "#E69F00", "#F0E442", "#0072B2", "#D55E00", "#009E73", "#999999")

# These are genes that are not included in the simulation and should be removed before comparing
# simulations with experimental data
IGNORE_GENES <- c("gene 10B",
                "possible gene 5.5-5.7",
                "gene 1.5",
                "gene 1.6",
                "gene 4.1",
                "gene 4B",
                "gene 0.6A",
                "gene 0.6B",
                "possible gene 0.6B",
                "gene 0.5",
                "gene 0.4")
```

```{r}
# Load labels
labels <- read_csv("output/id_map.csv") |> filter(!is.na(gene_number))
# Function to load and process simulation data
load_sim <- function(path, multiplier) {
  read_tsv(path,
          col_types = cols(
            species = col_character(),
            time = col_double(),
            protein = col_double(),
            transcript = col_double(),
            ribo_density = col_double()
          ),
          show_col_types = FALSE) |> 
    group_by(time) |> 
    nest() |>
    mutate(data = purrr::map(data, ~left_join(labels, ., by = c("pinetree" = "species")))) |>
    unnest(cols = c(data)) |>
    mutate(time = round(time)) |>
    mutate(protein = if_else(is.na(protein), 0, protein),
           transcript = if_else(is.na(transcript), 0, transcript),
           multiplier = multiplier) |>
    filter(!(pinetree %in% IGNORE_GENES), !is.na(gene_number))
}
```

```{r}
# Load simulation data for each multiplier
multipliers <- c(0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0)
sim_data <- lapply(multipliers, function(m) {
  file_path <- sprintf("data/simulation/phage/parallel_output/output_%.1f.tsv", m)
  load_sim(file_path, m)
})
sim_data <- bind_rows(sim_data)

max_time <- 1000

# Filter and select relevant data
sims <- sim_data |>
  filter(time == max_time) |>
  select(multiplier, protein) |>
  group_by(multiplier) |>
  summarize(protein = sum(protein))
sims

# Create the plot
plot <- ggplot(sims, aes(x = multiplier, y = protein)) +
  geom_line() +
  geom_point()+
  theme_minimal() +
  labs(x = "Weight Factor", 
    y = paste("Total Protein Abundance at", max_time, "s"), 
    title = paste("Total Protein Abundance at", max_time, "s\nfor Fractions of Optimal Codons   [ Using the Old Model ]")) +
  scale_x_continuous(breaks = seq(0.1, 1.0, by = 0.1), labels = seq(0.1, 1.0, by = 0.1))

if (SAVE_FIGS) {
  save_plot(sprintf("analysis/figures/old_total_proteins_by_weight_factor_%ds.pdf", max_time), plot, base_width = 6, base_height = 6)
}
```