---
title: "Fraction of Optimal Codons x Total Protein Abundance - Multiple Fractions of Preferred tRNAs"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)

setwd("/stor/work/Wilke/kelly/phage_simulation")

library(tidyverse)
library(ggplot2)
library(gtools)
library(stringr)
library(cowplot)
library(readxl)
SAVE_FIGS <- TRUE

cbPalette <- c("#CC79A7", "#56B4E9", "#E69F00", "#F0E442", "#0072B2", "#D55E00", "#009E73", "#999999")

# These are genes that are not included in the simulation and should be removed before comparing
# simulations with experimental data
IGNORE_GENES <- c("gene 10B",
                "possible gene 5.5-5.7",
                "gene 1.5",
                "gene 1.6",
                "gene 4.1",
                "gene 4B",
                "gene 0.6A",
                "gene 0.6B",
                "possible gene 0.6B",
                "gene 0.5",
                "gene 0.4")
```

```{r}
labels <- read_csv("output/id_map.csv") %>% filter(!is.na(gene_number))
load_sim <- function(paths) {
  data_list <- lapply(paths, function(path) {
    read_tsv(path) %>%
      mutate(file = path) %>%
      group_by(time) %>%
      nest() %>%
      mutate(data = purrr::map(data, ~full_join(labels, ., by = c("pinetree" = "species")))) %>%
      unnest() %>%
      mutate(time = round(time)) %>%
      mutate(protein = if_else(is.na(protein), 0, protein),
             transcript = if_else(is.na(transcript), 0, transcript)) %>%
      filter(!(pinetree %in% IGNORE_GENES), !is.na(gene_number))
  })
  bind_rows(data_list)
}

# Define the output directories and max time
output_dirs <- c("data/simulation/phage/codon_parallel_pref0.7_charge10",
                 "data/simulation/phage/codon_parallel_pref0.5_charge10")
max_time <- 825

# Extract pref_val and charge_val from the output directory string
pref_vals <- as.numeric(str_extract(output_dirs, "(?<=pref)[0-9]\\.[0-9]"))
charge_val <- as.numeric(str_extract(output_dirs[1], "(?<=charge)[0-9]+"))

# Generate file paths for each charge rate
seeds <- 1:3
fop_proportions <- 0:10

file_paths <- expand.grid(seed = seeds, fop = fop_proportions, pref_val = pref_vals) %>%
  mutate(output_dir = ifelse(pref_val == 0.7, output_dirs[1], output_dirs[2])) %>%
  mutate(file_name = sprintf("trna_phage_pref%.1f_%d_fop%d.tsv", pref_val, seed, fop)) %>%
  mutate(full_path = file.path(output_dir, file_name))

# Load all simulations
all_data <- load_sim(file_paths$full_path)

# Add seed, fop, pref columns
all_data <- all_data %>%
  mutate(
    seed = as.numeric(str_extract(file, "(?<=_)[0-9]+(?=_fop)")),
    fop = as.numeric(str_extract(file, "(?<=fop)[0-9]+")),
    pref = as.numeric(str_extract(file, "(?<=pref)[0-9]\\.[0-9]"))
  )

# Calculate total proteins for each proportion, charge, and seed at max time
total_proteins <- all_data %>%
  filter(time == max_time) %>%
  group_by(fop, pref, seed) %>%
  summarize(total_protein = sum(protein), .groups = "drop")
print(total_proteins)

# Calculate the average total proteins across seeds
avg_total_proteins <- total_proteins %>%
  group_by(fop, pref) %>%
  summarize(avg_total_protein = mean(total_protein), .groups = "drop")

# Create the plot
total_plot <- ggplot(avg_total_proteins, aes(x = fop, y = avg_total_protein, color = as.factor(pref))) +
  geom_line() +
  geom_point() +
  theme_minimal() +
  labs(x = "Fraction of Optimal Codons", 
    y = paste("Total Protein Abundance at", max_time, "s"), 
    title = paste("Total Protein Abundance at", max_time, "s\nfor Fractions of Optimal Codons   [ tRNA Charge Rate:", charge_val, "]"),
    color = "Fraction of Preferred tRNAs") +
  scale_x_continuous(breaks = seq(0, 10, by = 1), labels = seq(0, 1, by = 0.1)) +
  scale_color_manual(values = cbPalette) 

if (SAVE_FIGS) {
  save_plot(sprintf("analysis/figures/total_proteins_by_codon_proportion_%ds_multipref_charge%d.pdf", max_time, charge_val), total_plot, base_width = 6, base_height = 6)
}
```